### PR TITLE
[AR-4709] Create and Edit delegate

### PR DESCRIPTION
**Pull request by karthik-durai**
# Pull Request

Resolves  [AR-4709](https://team-1624093970686.atlassian.net/browse/AR-4709).

## Changes

1. Implemented logic and UI for creating delegate.
2.Implemented logic and UI for editing delegate.

## Checklist

- [X] The branch name follows the format: `developer/AR-XXX-issue-name`.
- [X] The changes have been tested locally.

### CODE REVIEW

File: src/components/app-configure/access/DeleteDelegate.vue [hunk 1]
```diff
@@ -0,0 +1,79 @@
+<script lang="ts" setup>
```
Comment (shrinathprabhu) [inline src/components/app-configure/access/DeleteDelegate.vue:1]: Remove this component. Is it used anywhere?

File: src/components/app-configure/access/CreateDelegate.vue [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+<template>
+  <VOverlay>
+    <div
+      style="
+        display: flex;
+        flex-direction: column;
+        align-items: center;
+        justify-content: center;
+        width: 100%;
+        height: 100%;
```
Comment (shrinathprabhu) [inline src/components/app-configure/access/CreateDelegate.vue:85]: Can we move this to `<style>` or if it's reused we can move it to `components/lib/styles/style-guide.css`

File: src/components/app-configure/access/DeleteDelegate.vue (hunk 1 above)
Comment (karthik-durai) [reply src/components/app-configure/access/DeleteDelegate.vue:1]: ok

File: src/components/app-configure/access/DeleteDelegate.vue [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+          <VStack justify="space-between" align="center">
+            <h1 class="sub-heading-5 font-700 text-uppercase">
+              Create A Delegate
+            </h1>
+            <Button
+              style="
+                cursor: pointer;
+                background-color: transparent;
+                border: none;
+              "
```
Comment (shrinathprabhu) [inline src/components/app-configure/access/DeleteDelegate.vue:34]: Move to `<style>` tag

File: src/components/app-configure/access/DeleteDelegate.vue [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  <VOverlay>
+    <div
+      style="
+        display: flex;
+        flex-direction: column;
+        align-items: center;
+        justify-content: center;
+        width: 100%;
+        height: 100%;
+      "
```
Comment (shrinathprabhu) [inline src/components/app-configure/access/DeleteDelegate.vue:21]: Move to `<style>` tag

File: src/components/app-configure/access/CreateDelegate.vue [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    toast.error('Please fill all required values')
+  }
+  showLoader('Created Delegate...')
+  try {
+    const payload = {
+      name: delegateName.value,
+      address: selectedDelegateAddress.value,
+      permissions: ['Dowload', 'Reshare'],
+    }
+    const { data } = await createDelegate(appId, payload)
```
Comment (shrinathprabhu) [inline src/components/app-configure/access/CreateDelegate.vue:55]: Need to perform smartcontract operation if this is successful

Need to call `grantDelegate` function from `smart-contract.service.ts`

Comment (karthik-durai) [reply src/components/app-configure/access/CreateDelegate.vue:55]: ok

File: src/services/smart-contract.service.ts [hunk 6]
```diff
@@ -24,6 +25,30 @@ async function setAppConfig(appName: string, socialAuth: AppConfigCred[]) {
...
 }
 
+async function enableUiMode() {
+  await signTransaction('setUiMode')
+}
+
+async function grantDelegate(
+  keyAddress: string,
+  permissions: DelegatePermission[],
+  appId: AppId
```
Comment (shrinathprabhu) [inline src/services/smart-contract.service.ts:35]: make this (appId, keyAddress, permissions) or turn this into object

File: src/components/app-configure/access/EditDelegate.vue [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    toast.error('Please fill all required values')
+  }
+  try {
+    const payload = {
+      id: props.delegateInfo.id,
+      name: delegateName.value,
+      address: selectedDelegateAddress.value,
+      permissions: ['Dowload', 'Reshare'],
+    }
+    const { data } = await editDelegate(appId, payload)
```
Comment (shrinathprabhu) [inline src/components/app-configure/access/EditDelegate.vue:48]: Need to perform smartcontract operation if this is successful

Need to revoke the old delegate address and grant the new delegate address

```js
await revokeDelegate(appId, oldAddress)
await grantDelegate(appId, newAddress, permissions)
```

File: src/components/app-configure/access/GenerateKeySuccess.vue [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+}
+</script>
+
+<template>
+  <VOverlay>
+    <div class="center height-100 width-100">
+      <VCard
+        popup="true"
+        class="flex column"
+        style="max-width: 600px; padding: 30px"
```
Comment (shrinathprabhu) [inline src/components/app-configure/access/GenerateKeySuccess.vue:38]: Move this to `<style>`

File: src/components/app-configure/access/CreateDelegate.vue [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+import { useAppId } from '@/use/getAppId'
+
+const emit = defineEmits(['close', 'generateKey', 'created'])
+
+const props = defineProps({
+  delegateKeys: {
+    type: Array,
+    required: true,
+  },
+})
```
Comment (ajithranka) [inline src/components/app-configure/access/CreateDelegate.vue:24]: Use TypeScript to define the prop types. Vue's object syntax for defining props is quite limited — `Array`, for example, could be an array of anything.

File: src/components/app-configure/access/CreateDelegate.vue [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+  loader.value.message = message
+}
+
+function hideLoader() {
+  loader.value.show = false
+  loader.value.message = ''
+}
+
+async function onSubmit() {
+  if (!delegateName.value.length || !selectedDelegateAddress.value.length) {
```
Comment (ajithranka) [inline src/components/app-configure/access/CreateDelegate.vue:47]: Prefer avoiding implicit coercion is conditionals.

```suggestion
  if (delegateName.value === "" || selectedDelegateAddress.value === "") {
```

File: src/components/app-configure/access/CreateDelegate.vue [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    toast.error('Please fill all required values')
+  }
+  showLoader('Created Delegate...')
```
Comment (ajithranka) [inline src/components/app-configure/access/CreateDelegate.vue]:
```suggestion
  showLoader('Creating delegate...')
```

File: src/components/app-configure/access/CreateDelegate.vue [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+        </VStack>
+      </VCard>
+    </div>
+  </VOverlay>
+</template>
+
+<style scoped>
+p {
+  margin: 0;
+}
```
Comment (ajithranka) [inline src/components/app-configure/access/CreateDelegate.vue:161]: This should be part of our reset styles.

File: src/components/app-configure/access/DidDelegate.vue [hunk 13]
```diff
@@ -20,14 +24,72 @@ const loaderStore = useLoaderStore()
 const toast = useToast()
 
 const app = appsStore.app(appId)
-const delegates = computed(() => app.access.delegates)
+const delegates = ref(app.access.delegates)
 
-const showDeletePopup = ref(false)
+type selectedOperation = 'create' | 'edit'
```
Comment (ajithranka) [inline src/components/app-configure/access/DidDelegate.vue:29]: I think we might have this type defined somewhere else too. 

```suggestion
type DelegateOperation = 'create' | 'edit'
```

File: src/components/app-configure/access/DidDelegate.vue [hunk 14]
```diff
@@ -20,14 +24,72 @@ const loaderStore = useLoaderStore()
...
+const showGenerateKeySuccess = ref(false)
+const showEditDelegate = ref(false)
+const showDeletePopup = ref(false)
+const delegateToEdit = ref({})
 const selectedDelegate: Ref<Delegate | null> = ref(null)
+let currentSelectedOperation: selectedOperation = 'edit'
+
+async function getDelegateKeys() {
+  try {
+    loaderStore.showLoader('Fethcing keys...')
```
Comment (ajithranka) [inline src/components/app-configure/access/DidDelegate.vue:43]:
```suggestion
    loaderStore.showLoader('Fetching keys...')
```

File: src/components/app-configure/access/DidDelegate.vue [hunk 15]
```diff
@@ -20,14 +24,72 @@ const loaderStore = useLoaderStore()
...
+  delegates.value = app.access.delegates
+}
+
+async function onEditingDelegate() {
+  await appsStore.fetchAndStoreAppConfig(appId)
+  const app = appsStore.app(appId)
+  delegates.value = app.access.delegates
+}
+
+async function onEditClick(delegate: object) {
```
Comment (ajithranka) [inline src/components/app-configure/access/DidDelegate.vue:81]: Can we use a more specific type for `delegate`?

File: src/components/app-configure/access/EditDelegate.vue (hunk 7 above)
Comment (karthik-durai) [reply src/components/app-configure/access/EditDelegate.vue:55]: https://github.com/arcana-network/developer-dashboard/pull/156/commits/7b6e188f7d1078eddc2f39cf7bc45158ec35f195

File: src/services/smart-contract.service.ts (hunk 6 above)
Comment (karthik-durai) [reply src/services/smart-contract.service.ts]: https://github.com/arcana-network/developer-dashboard/pull/156/commits/7b6e188f7d1078eddc2f39cf7bc45158ec35f195

File: src/components/app-configure/access/EditDelegate.vue [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+const props = defineProps({
+  delegateInfo: {
+    type: Object,
+    required: true,
+  },
+  delegateKeys: {
+    type: Array,
+    required: true,
+  },
+})
```
Comment (ajithranka) [inline src/components/app-configure/access/EditDelegate.vue:31]: Prefer using TypeScript for defining the prop types.

File: src/components/app-configure/access/DeleteDelegate.vue (hunk 1 above)
Comment (karthik-durai) [reply src/components/app-configure/access/DeleteDelegate.vue]: https://github.com/arcana-network/developer-dashboard/pull/156/commits/31946dcfa27636ce0d82d033445c2b6c9a907d3b

File: src/components/app-configure/access/EditDelegate.vue [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+async function onSubmit() {
+  if (!delegateName.value.length || !selectedDelegateAddress.value.length) {
+    toast.error('Please fill all required values')
+  }
+  try {
+    const payload = {
+      id: props.delegateInfo.id,
+      name: delegateName.value,
+      address: selectedDelegateAddress.value,
+      permissions: ['Dowload', 'Reshare'],
```
Comment (ajithranka) [inline src/components/app-configure/access/EditDelegate.vue]: Will this cause an error?

```suggestion
      permissions: ['Download', 'Reshare'],
```

File: src/components/app-configure/access/GenerateKeySuccess.vue (hunk 8 above)
Comment (karthik-durai) [reply src/components/app-configure/access/GenerateKeySuccess.vue:38]: You mean to `.css`

**Review by ajithranka — CHANGES_REQUESTED** [review]

File: src/components/app-configure/access/CreateDelegate.vue (hunk 11 above)
Comment (karthik-durai) [reply src/components/app-configure/access/CreateDelegate.vue]: done

File: src/components/app-configure/access/EditDelegate.vue (hunk 17 above)
Comment (karthik-durai) [reply src/components/app-configure/access/EditDelegate.vue]: Fixed it with help of Saurav

File: src/components/app-configure/access/CreateDelegate.vue (hunk 10 above)
Comment (karthik-durai) [reply src/components/app-configure/access/CreateDelegate.vue]: https://github.com/arcana-network/developer-dashboard/pull/156/commits/2981de3c4a5362ec0e344c13d12de4efb01075b4

File: src/components/app-configure/access/DidDelegate.vue (hunk 14 above)
Comment (karthik-durai) [reply src/components/app-configure/access/DidDelegate.vue]: https://github.com/arcana-network/developer-dashboard/pull/156/commits/8a4f203ecb70f08af921f8b9e96a9b4f4d997c8b

File: src/components/app-configure/access/CreateDelegate.vue (hunk 12 above)
Comment (karthik-durai) [reply src/components/app-configure/access/CreateDelegate.vue]: https://github.com/arcana-network/developer-dashboard/pull/156/commits/9f7e26f0d802ca2d5df75784932102835e0c9bc9

File: src/components/app-configure/access/DidDelegate.vue (hunk 15 above)
Comment (karthik-durai) [reply src/components/app-configure/access/DidDelegate.vue]: https://github.com/arcana-network/developer-dashboard/pull/156/commits/04175c6cb8d355df114957bc9ef3b852faae92f2

File: src/components/app-configure/access/EditDelegate.vue (hunk 16 above)
Comment (karthik-durai) [reply src/components/app-configure/access/EditDelegate.vue]: https://github.com/arcana-network/developer-dashboard/pull/156/commits/764ff0c48c67ed42721b0494b417f70842081ebb

File: src/components/app-configure/access/CreateDelegate.vue (hunk 9 above)
Comment (karthik-durai) [reply src/components/app-configure/access/CreateDelegate.vue]: https://github.com/arcana-network/developer-dashboard/pull/156/commits/764ff0c48c67ed42721b0494b417f70842081ebb

File: src/components/app-configure/access/GenerateKeySuccess.vue (hunk 8 above)
Comment (shrinathprabhu) [reply src/components/app-configure/access/GenerateKeySuccess.vue:38]: No, there is `<style scoped>` at the bottom, I guess we can move internal styles there

File: src/components/app-configure/access/CreateDelegate.vue (hunk 10 above)
Comment (shrinathprabhu) [reply src/components/app-configure/access/CreateDelegate.vue]: Should we also trim it?

File: src/services/smart-contract.service.ts [hunk 18]
```diff
@@ -24,6 +25,30 @@ async function setAppConfig(appName: string, socialAuth: AppConfigCred[]) {
   await signTransaction('setAppConfig', [appConfig])
 }
 
+async function enableUiMode() {
+  await signTransaction('setUiMode')
+}
```
[the remark this answers is not in the document]
Comment (karthik-durai) [reply src/services/smart-contract.service.ts]: https://github.com/arcana-network/developer-dashboard/pull/156/commits/7d9bc3ecccbc6e1954afc9201a53352af095876c

**Review by ajithranka — APPROVED** [review]

File: src/components/app-configure/access/GenerateKeySuccess.vue (hunk 8 above)
Comment (karthik-durai) [reply src/components/app-configure/access/GenerateKeySuccess.vue]: https://github.com/arcana-network/developer-dashboard/pull/156/commits/abd5da92c44f00b43b56e3b16c0b885131ed0a11

**Pull request merged**